solutions/ex2/Main.hs

```diff
@@ -389,4 +391,66 @@ instance MakeClearVariableStatement MJavaSig where
 
 
 exampleLuaProgram :: IO (Maybe (MLuaTerm LBlockL))
-exampleLuaProgram = parseFile @MLuaSig "input-files/lua/Foo.lua"
\ No newline at end of file
+exampleLuaProgram = parseFile @MLuaSig "input-files/lua/Foo.lua"
+
+-- Hints about supporting Lua:
+-- Create a new typeclass that says that `fs` can make Block from BlockItems somehow, not necessarily via SimpleBlock,
+-- and allowing the user to specify a non-standard "block end" term.
+-- Use UndecidableInstances as there are too many constraints.
```

Comment:
UndecidableInstances is not based on the number of constraints, but rather the shape.

I don't know the exact rules. Not many people do, because the rule "add UndecidableInstances when the compiler tells you to" works pretty well.